src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,388 @@
+}
+
+jsg::JsValue deserializeV8(jsg::Lock& js, kj::ArrayPtr<const kj::byte> ser) {
+  jsg::Deserializer deserializer(js, kj::mv(ser), kj::none, kj::none,
+      jsg::Deserializer::Options {
+    .version = 15,
+    .readHeader = true,
+  });
+
+  return deserializer.readValue(js);
+}
+} // namespace
+
+kj::Promise<capnp::Response<rpc::CallResults>> WorkerRpc::sendWorkerRpc(
+    jsg::Lock& js,
+    kj::StringPtr name,
+    const v8::FunctionCallbackInfo<v8::Value>& args) {
+
+  auto& ioContext = IoContext::current();
+  auto worker = getClient(ioContext, kj::none, "getJsRpcTarget"_kjc);
+  auto event = kj::heap<api::GetJsRpcTargetCustomEventImpl>(WORKER_RPC_EVENT_TYPE);
+
+  rpc::JsRpcTarget::Client client = event->getCap();
+  auto builder = client.callRequest();
+  builder.setMethodName(name);
+
+  kj::Vector<jsg::JsValue> argv(args.Length());
+  for (int n = 0; n < args.Length(); n++) {
+    argv.add(jsg::JsValue(args[n]));
+  }
+
+  // If we have arguments, serialize them.
+  // Note that we may fail to serialize some element, in which case this will throw back to JS.
+  if (argv.size() > 0) {
+    auto ser = serializeV8(js, js.arr(argv.asPtr()));
+    builder.initSerializedArgs().setV8Serialized(kj::mv(ser));
+  }
+
+  auto callResult = builder.send();
+  auto customEventResult = worker->customEvent(kj::mv(event));
+
+  co_return co_await callResult.attach(kj::mv(customEventResult));
+}
+
+jsg::Value WorkerRpc::handleRpcResponse(jsg::Lock& js, capnp::Response<rpc::CallResults> result) {
+  switch(result.which()) {
+    case rpc::CallResults::EXCEPTION: {
+      js.throwException(deserializeV8(js, result.getException().getV8Serialized().asBytes()));
+    }
+    case rpc::CallResults::VALUE: {
+      return jsg::Value(js.v8Isolate, deserializeV8(js,
+          result.getValue().getV8Serialized().asBytes()));
+    }
+  }
+}
+
 kj::Maybe<jsg::JsValue> WorkerRpc::getNamed(jsg::Lock& js, kj::StringPtr name) {
-    // TODO(soon): The server side PR will fill this function with the RPC call and other stuff,
-    //             this is just a temporary placeholder for testing.
-    return jsg::JsValue(js.wrapReturningFunction(js.v8Context(),
-        [](jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
-          return js.error("WorkerRpc is unimplemented");
+  // Named intercept is enabled, this means we won't default to legacy behavior.
+  // The return value of the function is a promise that resolves once the remote returns the result
+  // of the RPC call.
+  return jsg::JsValue(js.wrapReturningFunction(js.v8Context(), [this, methodName=kj::str(name)]
+      (jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+        auto& ioContext = IoContext::current();
+        // Wait for the RPC to resolve and then process the result.
+        return js.wrapSimplePromise(ioContext.awaitIo(js, sendWorkerRpc(js, methodName, args),
+            [this] (jsg::Lock& js, capnp::Response<rpc::CallResults> result) -> jsg::Value {
```

Comment:
You'll want to ensure that a strong reference to `this` is maintained for this to be safe. That said, there's nothing in `handleRpcResponse(...)` that appears to rely on the internal state of `this` so maybe that can be just a local function in an anonymous namespace rather than a function hung off `WorkerRpc`? Then you wouldn't need to capture anything here.

Reply:
Moved into the anonymous namespace higher up in the file and dropped the `this` capture.